### ISSUE TITLE
Causal masking with past

### ISSUE THREAD

**Issue by turboderp** [Dao-AILab/flash-attention#351]
Is there a way to use `flash_attn_func` with a causal mask that isn't square? Currently it seems like only two modes are supported for causal LM: either processing the whole input sequence from the beginning (i.e. without a cached past), or individual tokens (sequence length of 1 with no causal mask.)

Ideally I'd want it to apply a mask like this:

    1 1 1 1 1 1 0 0 0
    1 1 1 1 1 1 1 0 0
    1 1 1 1 1 1 1 1 0
    1 1 1 1 1 1 1 1 1

Which would make sense for say adding four tokens to a five-token sequence. But as far as I can tell Flash Attention applies a mask like:

    1 0 0 0 0 0 0 0 0
    1 1 0 0 0 0 0 0 0
    1 1 1 0 0 0 0 0 0
    1 1 1 1 0 0 0 0 0

And I can't understand when that would be useful. It does appear to be the same behavior as `scaled_dot_product_attention` from PyTorch and xformers, so I guess it's some kind of standard? Torch at least lets you specify an arbitrary mask. Does Flash Attention have an option for this?

**Comment by jlamypoirier** [comment]
I would also like that, and agree the default upper triangular version seems useless. The causal thing is usually unnecessary with a past because sq is usually 1 but there are cases when we do want it.

I'm actually trying to implement this and I think I got at least the forward pass working. It's really just about changing the mask from `col<row` to `col-sk<row-sq` https://github.com/ServiceNow/flash-attention/pull/1/files#diff-0f0d113fa6e3f51dffc22ef0b84f6a0a50a5410ce33549b8986da9afe4b7f08bL156, but there are complications in the code to discard whole blocks https://github.com/ServiceNow/flash-attention/pull/1/files#diff-9e1775131ae22a74dc4e0333c57539573394a059db8097ef74fae24243347ce1L145.

**Comment by tridao** [comment]
This behavior might have been an accident of history. When I was implementing `is_causal` I only thought of the self-attention case where Q & K have the same length. For simplicity of implementation I chose to use the same indices comparison in the case where Q & K have different lengths.

Maybe at some point we should have an optional where the causal mask is "padded to the right" as mentioned here. Is there a good name for this flag? I'd prefer to keep `is_causal` as is for backward compatibility.

**Comment by jlamypoirier** [comment]
A flag would be fine, but is there even a case where we want the current behavior?

**Comment by tridao** [comment]
> A flag would be fine, but is there even a case where we want the current behavior?

I've no idea if anyone is relying on the current behavior when seqlen_q != seqlen_k.
Also it's not clear what the causal mask should look like if seqlen_q > seqlen_k.

**Comment by jlamypoirier** [comment]
I think anything else than what is suggested in this issue would be weird since it would mean the key has been computed for the future and we're just willingly wasting that computation.

There might be one potential use case though, where one would pre-allocate the key-value cache and assume the non-computed future values are ignored. (I'm also trying to do that but looks like all that's needed is overriding the seqlen_k variable)

**Comment by turboderp** [comment]
>The causal thing is usually unnecessary with a past because sq is usually 1 but there are cases when we do want it.

Well, specifically I would use it in two cases:

1. To run inference on very long sequences, in chunks. This is a common way to get around the quadratic scaling of regular attention.
2. In a back-and forth chatbot style application where a context is built either token by token (generated responses) or in larger chunks (user inputs), without having to re-compute all keys and values for every interaction.

It's always baffled me that Torch SDP and xformers use a "left-triangular" mask as they do, since I don't think those two are particularly strange or niche use cases. I guess \#1 is less relevant for FA, of course.

**Comment by tridao** [comment]
This is great. I didn't think of these use cases before.
Yeah I'm inclined to change the behavior of `is_causal`. This will take some time to implement (I'm focusing on fixing some edge cases with the backward pass right now).
I'll also need to need to (1) make sure it doesn't affect performance in the common use case such as GPT-style LM training (2) check with some folks that depend FlashAttention about breaking backward compatibility.

**Comment by BoxiangW** [comment]
I tried to implement this without changing `is_causal`, it would be great if anyone would like to check. https://github.com/Dao-AILab/flash-attention/pull/436

**Comment by tridao** [comment]
This is now implemented.
